File: strategy_core/http.py
"""Engine-owned HTTP request/response interfaces for strategy code."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Literal, Protocol, runtime_checkable

if TYPE_CHECKING:
    from strategy_core.models import JSONValue

HttpMethod = Literal["GET", "POST", "PUT", "PATCH", "DELETE"]
HttpHeaders = Mapping[str, str]
HttpParams = Mapping[str, str | int | float | bool | None]


@dataclass(frozen=True, slots=True)
class HttpRequest:
    """Normalized outbound HTTP request shape."""

    method: HttpMethod
    url: str
    headers: dict[str, str] = field(default_factory=dict)
    params: dict[str, str | int | float | bool | None] = field(default_factory=dict)
    json_body: JSONValue | None = None
    text_body: str | None = None
    timeout_seconds: float | None = None


@dataclass(frozen=True, slots=True)
class HttpResponse:
    """Normalized HTTP response shape returned to strategies."""

    status_code: int
    headers: dict[str, str] = field(default_factory=dict)
    text: str | None = None
    json_body: JSONValue | None = None


@runtime_checkable
class HttpClient(Protocol):
    """Engine-managed HTTP surface so runtimes can observe and control requests."""

    async def request(self, request: HttpRequest) -> HttpResponse: ...

    async def get(
        self,
        url: str,
        *,
        headers: HttpHeaders | None = None,
        params: HttpParams | None = None,
        timeout_seconds: float | None = None,
    ) -> HttpResponse: ...

    async def post(
        self,
        url: str,
        *,
        headers: HttpHeaders | None = None,
        params: HttpParams | None = None,
        json_body: JSONValue | None = None,
        text_body: str | None = None,
        timeout_seconds: float | None = None,
    ) -> HttpResponse: ...

File: native/strategy_core/src/http.rs
use std::collections::BTreeMap;
use std::future::Future;

use serde::{Deserialize, Serialize};

use crate::models::JsonValue;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Patch,
    Delete,
}

pub type HttpHeaders = BTreeMap<String, String>;
pub type HttpParams = BTreeMap<String, JsonValue>;

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct HttpRequest {
    pub method: HttpMethod,
    pub url: String,
    #[serde(default)]
    pub headers: HttpHeaders,
    #[serde(default)]
    pub params: HttpParams,
    pub json_body: Option<JsonValue>,
    pub text_body: Option<String>,
    pub timeout_seconds: Option<f64>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct HttpResponse {
    pub status_code: i64,
    #[serde(default)]
    pub headers: HttpHeaders,
    pub text: Option<String>,
    pub json_body: Option<JsonValue>,
}

pub trait HttpClient {
    type Error;

    fn request(
        &self,
        request: HttpRequest,
    ) -> impl Future<Output = Result<HttpResponse, Self::Error>> + Send;

    fn get(
        &self,
        url: &str,
        headers: Option<HttpHeaders>,
        params: Option<HttpParams>,
        timeout_seconds: Option<f64>,
    ) -> impl Future<Output = Result<HttpResponse, Self::Error>> + Send;

    fn post(
        &self,
        url: &str,
        headers: Option<HttpHeaders>,
        params: Option<HttpParams>,
        json_body: Option<JsonValue>,
        text_body: Option<String>,
        timeout_seconds: Option<f64>,
    ) -> impl Future<Output = Result<HttpResponse, Self::Error>> + Send;
}
